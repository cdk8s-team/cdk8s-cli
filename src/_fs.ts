import { writeFileSync } from 'fs-extra';

export const _fs = {
  writeFileSync: writeFileSync,
};