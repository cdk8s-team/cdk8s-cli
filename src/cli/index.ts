import { yellow, red } from 'colors';
import * as yargs from 'yargs';
import { upgradeAvailable } from '../upgrades';

if (new Date() < new Date('2025-01-01')) {
  console.warn(yellow('⚠️ Version 1.x of the cdk8s toolchain is deprecated and will become end-of-life on 01/01/25.'));
  console.warn(yellow('We strongly recommend migrating to 2.x. See https://cdk8s.io/docs/latest/migrating-from-1.x/'));
} else {
  console.warn(red('⚠️ Version 1.x of the cdk8s toolchain is end-of-life (since 01/01/25).'));
  console.warn(red('We strongly recommend migrating to 2.x. See https://cdk8s.io/docs/latest/migrating-from-1.x/'));
}

async function main() {
  const ya = yargs
    .option('check-upgrade', { type: 'boolean', desc: 'Check for cdk8s-cli upgrade', default: true })
    .check(argv => {
      if (argv.checkUpgrade) {
        const versions = upgradeAvailable();
        if (versions) {
          console.error('------------------------------------------------------------------------------------------------');
          console.error(yellow(`A new version ${versions.latest} of cdk8s-cli is available (current ${versions.current}).`));
          console.error(yellow('Run "npm install -g cdk8s-cli" to install the latest version on your system.'));
          console.error(yellow('For additional installation methods, see https://cdk8s.io/docs/latest/getting-started'));
          console.error('------------------------------------------------------------------------------------------------');
        }
      }

      return true;
    }, true)
    .commandDir('cmds')
    .recommendCommands()
    .wrap(yargs.terminalWidth())
    .showHelpOnFail(false)
    .env('CDK8S')
    .epilogue('Options can be specified via environment variables with the "CDK8S_" prefix (e.g. "CDK8S_OUTPUT")')
    .help();

  const args = ya.argv;
  if (args._.length === 0) {
    yargs.showHelp();
  }
}

main().catch(e => {
  console.error(e.stack);
  process.exit(1);
});
