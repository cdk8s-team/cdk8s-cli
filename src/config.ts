import * as fs from 'fs-extra';
import * as yaml from 'yaml';
import { Language } from './import/base';

const CONFIG_FILE = 'cdk8s.yaml';

export interface ImportSpec {
  readonly moduleNamePrefix?: string;
  readonly source: string;
}

export interface ValidationConfig {
  readonly package: string;
  readonly version: string;
  readonly class: string;
  readonly installEnv?: { [key: string]: any };
  readonly properties?: { [key: string]: any };
}

export enum SynthesisFormat {
  PLAIN = 'plain',
  HELM = 'helm',
}

export enum HelmChartApiVersion {
  V1 = 'v1',
  V2 = 'v2',
}

export interface SynthConfig {
  readonly format?: SynthesisFormat;
  readonly chartApiVersion?: HelmChartApiVersion;
  readonly chartVersion?: string;
}

export interface Config {
  readonly app?: string;
  readonly language?: Language;
  readonly output?: string;
  readonly imports?: string[];
  readonly pluginsDirectory?: string;
  readonly validations?: string | ValidationConfig[];
  readonly synthConfig?: SynthConfig;
}

export function readConfigSync(): Config | undefined {
  if (fs.existsSync(CONFIG_FILE)) {
    const config = {
      ...yaml.parse(fs.readFileSync(CONFIG_FILE, 'utf-8')),
    };

    return config;
  }

  return undefined;
}

export async function addImportToConfig(source: string) {
  let curConfig = yaml.parse(fs.readFileSync(CONFIG_FILE, 'utf-8'));

  const curImports = curConfig.imports ?? [];
  if (!curImports.includes(source)) {
    const importsList = curConfig.imports ?? [];
    importsList.push(source);
    let config = {
      ...curConfig,
      imports: importsList,
    };
    await fs.outputFile(CONFIG_FILE, yaml.stringify(config));
  }
}