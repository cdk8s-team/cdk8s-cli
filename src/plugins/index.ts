export { Validation, ValidationContext, ValidationViolation, ValidationViolatingResource } from './validation';