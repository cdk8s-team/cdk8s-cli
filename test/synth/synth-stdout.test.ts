import * as path from 'path';
import * as fs from 'fs-extra';
import * as yaml from 'yaml';
import { Config, ValidationConfig } from '../../src/config';
import { findConstructMetadata, mkdtemp } from '../../src/util';

beforeEach(() => {
  // resetting so that every test can use a different config file,
  // which is read on module load.
  jest.resetModules();
});

test('synth with both --stdout and --output throws exception', () => {

  const cmd = requireSynth();

  // eslint-disable-next-line
  expect(cmd.handler({ app: 'cdk8s', output: 'test', stdout: true })).rejects.toEqual(new Error('\'--output\' and \'--stdout\' are mutually exclusive. Please only use one.'));
});

describe('validations', () => {

  test('synth with inline validations', async () => {

    const validations: ValidationConfig[] = [{
      package: path.join(__dirname, '__resources__', 'validation-plugin'),
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        fail: false,
      },
    }];

    await synth({ validations });
  });

  test('synth with local validations file', async () => {

    const validationsFile = './validations.yaml';

    const validations: ValidationConfig[] = [{
      package: path.join(__dirname, '__resources__', 'validation-plugin'),
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        fail: false,
      },
    }];

    await synth({
      validations: validationsFile,
      preSynth: async (dir: string) => {
        fs.writeFileSync(path.join(dir, validationsFile), yaml.stringify(validations));
      },
    });

  });

  test('synth with validation plugin specified as relative path', async () => {

    const validations: ValidationConfig[] = [{
      package: './validation-plugin',
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        fail: false,
      },
    }];

    await synth({
      validations,
      preSynth: async (dir: string) => {
        fs.copySync(path.join(__dirname, '__resources__', 'validation-plugin'), path.join(dir, 'validation-plugin'));
      },
    });

  });

  test('synth with validation plugin specified as absolute path', async () => {

    const validations: ValidationConfig[] = [{
      package: path.join(__dirname, '__resources__', 'validation-plugin'),
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        fail: false,
      },
    }];

    await synth({ validations });

  });

  test('synth fails when validation plugin specified as url', async () => {

    const validations: ValidationConfig[] = [{
      package: 'http://path/to/plugin',
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        fail: false,
      },
    }];

    await expect(() => synth({ validations })).rejects.toThrow(/Unsupported package reference/);

  });

  test('synth fails when validation reports failure', async () => {

    const validations: ValidationConfig[] = [{
      package: path.join(__dirname, '__resources__', 'validation-plugin'),
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        fail: true,
      },
    }];

    await expect(() => synth({ validations })).rejects.toThrow('Code: 2');

  });

  test('synth skips validation if no-validate is passed', async () => {

    const validations: ValidationConfig[] = [{
      package: path.join(__dirname, '__resources__', 'validation-plugin'),
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        fail: true,
      },
    }];

    await synth({ validations, validate: false });
  });

  test('synth fails when validations specify non existing local plugin', async () => {

    const plugin = path.join(__dirname, '__resources__', 'non-existing');
    const validations: ValidationConfig[] = [{
      package: plugin,
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        fail: true,
      },
    }];

    await expect(() => synth({ validations })).rejects.toThrow(/Cannot find module/);

  });

  test('synth fails when validations specify plugin version range', async () => {

    const validations: ValidationConfig[] = [{
      package: path.join(__dirname, '__resources__', 'validation-plugin'),
      version: '^1.0.0',
      class: 'MockValidation',
      properties: {
        fail: true,
      },
    }];

    await expect(() => synth({ validations })).rejects.toThrow(/Unsupported version spec/);

  });

  test('synth fails when validation plugin throws', async () => {

    const validations: ValidationConfig[] = [{
      package: path.join(__dirname, '__resources__', 'validation-plugin'),
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        throw: true,
      },
    }];

    await expect(() => synth({ validations })).rejects.toThrow(/Throwing per request/);

  });

  test('synth can write the validation reports to a file', async () => {

    const pluginPath = path.join(__dirname, '__resources__', 'validation-plugin');
    const dirNameRegex = new RegExp(__dirname, 'g');
    const validations: ValidationConfig[] = [{
      package: pluginPath,
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        fail: false,
      },
    }];
    await synth({
      validations,
      reportsFile: 'reports.json',
      postSynth: async (dir: string) => {
        const reports = fs.readFileSync(path.join(dir, 'reports.json'), { encoding: 'utf-8' })
          // '__dirname' contains environment specific paths, so it needs to be sanitized for consistent
          // results.
          .replace(dirNameRegex, '<__dirname-replaced>');
        expect(reports).toMatchSnapshot();
      },
    });

  });

  test('construct metadata is recorded by default when there are validations', async () => {

    const pluginPath = path.join(__dirname, '__resources__', 'validation-plugin');
    const validations: ValidationConfig[] = [{
      package: pluginPath,
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        fail: false,
      },
    }];
    await synth({
      validations,
      postSynth: async (dir: string) => {
        expect(findConstructMetadata(path.join(dir, 'dist/'))).toContain('construct-metadata.json');
      },
    });
  });

  test('construct metadata is NOT recorded by default when validations is empty', async () => {

    const validations: ValidationConfig[] = [];
    await synth({
      validations,
      postSynth: async (dir: string) => {
        expect(findConstructMetadata(path.join(dir, 'dist/'))).toBeUndefined();
      },
    });
  });

  test('construct metadata is NOT recorded by default when validations is undefined', async () => {

    const validations = undefined;
    await synth({
      validations,
      postSynth: async (dir: string) => {
        expect(findConstructMetadata(path.join(dir, 'dist/'))).toBeUndefined();
      },
    });
  });

  test('synth will not write the validation reports to an existing file', async () => {

    const pluginPath = path.join(__dirname, '__resources__', 'validation-plugin');
    const dirNameRegex = new RegExp(__dirname, 'g');
    const validations: ValidationConfig[] = [{
      package: pluginPath,
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        fail: false,
      },
    }];

    await expect(async () => {

      await synth({
        validations,
        reportsFile: 'reports.json',
        preSynth: async (dir: string) => {
          fs.writeFileSync(path.join(dir, 'reports.json'), 'hello');
        },
        postSynth: async (dir: string) => {
          const reports = fs.readFileSync(path.join(dir, 'reports.json'), { encoding: 'utf-8' })
            // '__dirname' contains environment specific paths, so it needs to be sanitized for consistent
            // results.
            .replace(dirNameRegex, '<__dirname-replaced>');
          expect(reports).toMatchSnapshot();
        },
      });

    }).rejects.toThrow(/Unable to write validation reports file. Already exists:/);

  });

  test('can pass environment to installation command', async () => {

    const validations: ValidationConfig[] = [{
      package: 'some-plugin',
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        throw: true,
      },
      installEnv: {
        // this should fail synth
        npm_config_registry: 'localhost:1234',
      },
    }];

    // to cover both node18 and below, which have different error messages
    // in this case.
    const re = new RegExp('Must be a full url|ERR_INVALID_URL');
    await expect(() => synth({ validations })).rejects.toThrow(re);

  });

  test('synth executed with --stdout', async () => {

    const validations: ValidationConfig[] = [{
      package: path.join(__dirname, '__resources__', 'validation-plugin'),
      version: '0.0.0',
      class: 'MockValidation',
      properties: {
        fail: false,
      },
    }];

    await synth({ validations, stdout: true });

  });

});

interface SynthOptions {

  readonly validations?: string | ValidationConfig[];
  readonly validate?: boolean;
  readonly stdout?: boolean;
  readonly reportsFile?: string;
  readonly preSynth?: (dir: string) => Promise<void>;
  readonly postSynth?: (dir: string) => Promise<void>;

}

async function synth(options: SynthOptions) {

  const app = `
const cdk8s = require('${require.resolve('cdk8s')}');
const app = new cdk8s.App();
const chart = new cdk8s.Chart(app, 'Chart');
new cdk8s.ApiObject(chart, 'Object', {
  kind: 'ConfigMap',
  apiVersion: 'v1',
  metadata: {
    name: 'config-map',
  }
});
app.synth();
`;

  await mkdtemp(async (dir: string) => {

    const stdout = options.stdout ?? false;
    const validate = options.validate ?? true;

    const config: Config = {
      validations: options.validations,
      app: 'node index.js',
      output: stdout ? undefined : 'dist',
      pluginsDirectory: path.join(dir, '.cdk8s', 'plugins'),
    };

    fs.writeFileSync(path.join(dir, 'index.js'), app);
    fs.writeFileSync(path.join(dir, 'cdk8s.yaml'), yaml.stringify(config));

    const recordConstructMetadata = !(options.validations == undefined || options.validations.length == 0);

    const pwd = process.cwd();
    const exit = process.exit;
    try {
      process.chdir(dir);
      process.env.CDK8S_RECORD_CONSTRUCT_METADATA = recordConstructMetadata ? 'true' : 'false';
      // our implementation does process.exit(2) so we need
      // to monkey patch it so we can assert on it.
      (process as any).exit = (code: number) => {
        throw new Error(`Code: ${code}`);
      };

      const cmd = requireSynth();
      if (options.preSynth) {
        await options.preSynth(dir);
      }
      await cmd.handler({
        app: config.app,
        output: config.output,
        stdout: stdout,
        validate: validate,
        pluginsDir: config.pluginsDirectory,
        validationReportsOutputFile: options.reportsFile ? path.join(dir, options.reportsFile) : undefined,
      });
      if (options.postSynth) {
        await options.postSynth(dir);
      }
      if (validate && findConstructMetadata(path.join(dir, 'dist/'))) {
        // this file is written by our test plugin
        const marker = path.join(dir, 'validation-done.marker');
        expect(fs.existsSync(marker)).toBeTruthy();
      }
    } finally {
      process.chdir(pwd);
      (process as any).exit = exit;
      delete process.env.CDK8S_RECORD_CONSTRUCT_METADATA;
    }

  });

}

function requireSynth() {
  const module = '../../src/cli/cmds/synth';
  // eslint-disable-next-line
  return require(module);
}
