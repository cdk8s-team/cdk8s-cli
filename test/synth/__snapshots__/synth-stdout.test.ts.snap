// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helm synthesis --chart-api-version is v1 with all inputs from cli and no config file is present 1`] = `
Object {
  "Chart.yaml": "apiVersion: v1
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis --chart-api-version is v1 with all inputs from config file and no related cli inputs 1`] = `
Object {
  "Chart.yaml": "apiVersion: v1
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis --chart-api-version is v1 with different inputs in cli and config file 1`] = `
Object {
  "Chart.yaml": "apiVersion: v1
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis --chart-api-version is v1 with inputs duplicated in cli and config file 1`] = `
Object {
  "Chart.yaml": "apiVersion: v1
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis --chart-api-version is v2 and crds are present with all inputs from cli and no config file is present 1`] = `
Object {
  "Chart.yaml": "apiVersion: v2
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "crds/__omitted__": "# Example from: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/

apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  # name must match the spec fields below, and be in the form: <plural>.<group>
  name: crontabs.stable.example.com
spec:
  # group name to use for REST API: /apis/<group>/<version>
  group: stable.example.com
  # list of versions supported by this CustomResourceDefinition
  versions:
    - name: v1
      # Each version can be enabled/disabled by Served flag.
      served: true
      # One and only one version must be marked as the storage version.
      storage: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                cronSpec:
                  type: string
                image:
                  type: string
                replicas:
                  type: integer
  # either Namespaced or Cluster
  scope: Namespaced
  names:
    # plural name to be used in the URL: /apis/<group>/<version>/<plural>
    plural: crontabs
    # singular name to be used as an alias on the CLI and for display
    singular: crontab
    # kind is normally the CamelCased singular type. Your resource manifests use this.
    kind: CronTab
    # shortNames allow shorter string to match your resource on the CLI
    shortNames:
    - ct",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis --chart-api-version is v2 and crds are present with all inputs from config file and no related cli inputs 1`] = `
Object {
  "Chart.yaml": "apiVersion: v2
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "crds/__omitted__": "# Example from: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/

apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  # name must match the spec fields below, and be in the form: <plural>.<group>
  name: crontabs.stable.example.com
spec:
  # group name to use for REST API: /apis/<group>/<version>
  group: stable.example.com
  # list of versions supported by this CustomResourceDefinition
  versions:
    - name: v1
      # Each version can be enabled/disabled by Served flag.
      served: true
      # One and only one version must be marked as the storage version.
      storage: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                cronSpec:
                  type: string
                image:
                  type: string
                replicas:
                  type: integer
  # either Namespaced or Cluster
  scope: Namespaced
  names:
    # plural name to be used in the URL: /apis/<group>/<version>/<plural>
    plural: crontabs
    # singular name to be used as an alias on the CLI and for display
    singular: crontab
    # kind is normally the CamelCased singular type. Your resource manifests use this.
    kind: CronTab
    # shortNames allow shorter string to match your resource on the CLI
    shortNames:
    - ct",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis --chart-api-version is v2 and crds are present with different inputs in cli and config file 1`] = `
Object {
  "Chart.yaml": "apiVersion: v2
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "crds/__omitted__": "# Example from: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/

apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  # name must match the spec fields below, and be in the form: <plural>.<group>
  name: crontabs.stable.example.com
spec:
  # group name to use for REST API: /apis/<group>/<version>
  group: stable.example.com
  # list of versions supported by this CustomResourceDefinition
  versions:
    - name: v1
      # Each version can be enabled/disabled by Served flag.
      served: true
      # One and only one version must be marked as the storage version.
      storage: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                cronSpec:
                  type: string
                image:
                  type: string
                replicas:
                  type: integer
  # either Namespaced or Cluster
  scope: Namespaced
  names:
    # plural name to be used in the URL: /apis/<group>/<version>/<plural>
    plural: crontabs
    # singular name to be used as an alias on the CLI and for display
    singular: crontab
    # kind is normally the CamelCased singular type. Your resource manifests use this.
    kind: CronTab
    # shortNames allow shorter string to match your resource on the CLI
    shortNames:
    - ct",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis --chart-api-version is v2 and crds are present with inputs duplicated in cli and config file 1`] = `
Object {
  "Chart.yaml": "apiVersion: v2
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "crds/__omitted__": "# Example from: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/

apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  # name must match the spec fields below, and be in the form: <plural>.<group>
  name: crontabs.stable.example.com
spec:
  # group name to use for REST API: /apis/<group>/<version>
  group: stable.example.com
  # list of versions supported by this CustomResourceDefinition
  versions:
    - name: v1
      # Each version can be enabled/disabled by Served flag.
      served: true
      # One and only one version must be marked as the storage version.
      storage: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                cronSpec:
                  type: string
                image:
                  type: string
                replicas:
                  type: integer
  # either Namespaced or Cluster
  scope: Namespaced
  names:
    # plural name to be used in the URL: /apis/<group>/<version>/<plural>
    plural: crontabs
    # singular name to be used as an alias on the CLI and for display
    singular: crontab
    # kind is normally the CamelCased singular type. Your resource manifests use this.
    kind: CronTab
    # shortNames allow shorter string to match your resource on the CLI
    shortNames:
    - ct",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis --chart-api-version is v2 without crds with all inputs from cli and no config file is present 1`] = `
Object {
  "Chart.yaml": "apiVersion: v2
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis --chart-api-version is v2 without crds with all inputs from config file and no related cli inputs 1`] = `
Object {
  "Chart.yaml": "apiVersion: v2
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis --chart-api-version is v2 without crds with different inputs in cli and config file 1`] = `
Object {
  "Chart.yaml": "apiVersion: v2
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis --chart-api-version is v2 without crds with inputs duplicated in cli and config file 1`] = `
Object {
  "Chart.yaml": "apiVersion: v2
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis filename url hash remains the same across synthesis with all inputs from cli and no config file is present 1`] = `
Object {
  "Chart.yaml": "apiVersion: v2
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "crds/__omitted__": "{
  \\"apiVersion\\": \\"apiextensions.k8s.io/v1\\",
  \\"kind\\": \\"CustomResourceDefinition\\",
  \\"metadata\\": {
    \\"name\\": \\"crontabs.stable.example.com\\"
  },
  \\"spec\\": {
    \\"group\\": \\"stable.example.com\\",
    \\"versions\\": [
      {
        \\"name\\": \\"v1\\",
        \\"served\\": true,
        \\"storage\\": true,
        \\"schema\\": {
          \\"openAPIV3Schema\\": {
            \\"type\\": \\"object\\",
            \\"properties\\": {
              \\"spec\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"cronSpec\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"image\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"replicas\\": {
                    \\"type\\": \\"integer\\"
                  }
                }
              }
            }
          }
        }
      }
    ],
    \\"scope\\": \\"Namespaced\\",
    \\"names\\": {
      \\"plural\\": \\"crontabs\\",
      \\"singular\\": \\"crontab\\",
      \\"kind\\": \\"CronTab\\",
      \\"shortNames\\": [
        \\"ct\\"
      ]
    }
  }
}",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis filename url hash remains the same across synthesis with all inputs from config file and no related cli inputs 1`] = `
Object {
  "Chart.yaml": "apiVersion: v2
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "crds/__omitted__": "{
  \\"apiVersion\\": \\"apiextensions.k8s.io/v1\\",
  \\"kind\\": \\"CustomResourceDefinition\\",
  \\"metadata\\": {
    \\"name\\": \\"crontabs.stable.example.com\\"
  },
  \\"spec\\": {
    \\"group\\": \\"stable.example.com\\",
    \\"versions\\": [
      {
        \\"name\\": \\"v1\\",
        \\"served\\": true,
        \\"storage\\": true,
        \\"schema\\": {
          \\"openAPIV3Schema\\": {
            \\"type\\": \\"object\\",
            \\"properties\\": {
              \\"spec\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"cronSpec\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"image\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"replicas\\": {
                    \\"type\\": \\"integer\\"
                  }
                }
              }
            }
          }
        }
      }
    ],
    \\"scope\\": \\"Namespaced\\",
    \\"names\\": {
      \\"plural\\": \\"crontabs\\",
      \\"singular\\": \\"crontab\\",
      \\"kind\\": \\"CronTab\\",
      \\"shortNames\\": [
        \\"ct\\"
      ]
    }
  }
}",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis filename url hash remains the same across synthesis with different inputs in cli and config file 1`] = `
Object {
  "Chart.yaml": "apiVersion: v2
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "crds/__omitted__": "{
  \\"apiVersion\\": \\"apiextensions.k8s.io/v1\\",
  \\"kind\\": \\"CustomResourceDefinition\\",
  \\"metadata\\": {
    \\"name\\": \\"crontabs.stable.example.com\\"
  },
  \\"spec\\": {
    \\"group\\": \\"stable.example.com\\",
    \\"versions\\": [
      {
        \\"name\\": \\"v1\\",
        \\"served\\": true,
        \\"storage\\": true,
        \\"schema\\": {
          \\"openAPIV3Schema\\": {
            \\"type\\": \\"object\\",
            \\"properties\\": {
              \\"spec\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"cronSpec\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"image\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"replicas\\": {
                    \\"type\\": \\"integer\\"
                  }
                }
              }
            }
          }
        }
      }
    ],
    \\"scope\\": \\"Namespaced\\",
    \\"names\\": {
      \\"plural\\": \\"crontabs\\",
      \\"singular\\": \\"crontab\\",
      \\"kind\\": \\"CronTab\\",
      \\"shortNames\\": [
        \\"ct\\"
      ]
    }
  }
}",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`Helm synthesis filename url hash remains the same across synthesis with inputs duplicated in cli and config file 1`] = `
Object {
  "Chart.yaml": "apiVersion: v2
name: __omitted__
version: 1.1.1
description: __omitted__
type: application
",
  "README.md": "This Helm chart is generated using cdk8s. Any manual changes to the chart would be discarded once cdk8s app is synthesized again with \`--format helm\`.",
  "crds/__omitted__": "{
  \\"apiVersion\\": \\"apiextensions.k8s.io/v1\\",
  \\"kind\\": \\"CustomResourceDefinition\\",
  \\"metadata\\": {
    \\"name\\": \\"crontabs.stable.example.com\\"
  },
  \\"spec\\": {
    \\"group\\": \\"stable.example.com\\",
    \\"versions\\": [
      {
        \\"name\\": \\"v1\\",
        \\"served\\": true,
        \\"storage\\": true,
        \\"schema\\": {
          \\"openAPIV3Schema\\": {
            \\"type\\": \\"object\\",
            \\"properties\\": {
              \\"spec\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"cronSpec\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"image\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"replicas\\": {
                    \\"type\\": \\"integer\\"
                  }
                }
              }
            }
          }
        }
      }
    ],
    \\"scope\\": \\"Namespaced\\",
    \\"names\\": {
      \\"plural\\": \\"crontabs\\",
      \\"singular\\": \\"crontab\\",
      \\"kind\\": \\"CronTab\\",
      \\"shortNames\\": [
        \\"ct\\"
      ]
    }
  }
}",
  "templates/chart-c86185a7.k8s.yaml": "apiVersion: v1
kind: ConfigMap
metadata:
  name: config-map
",
}
`;

exports[`validations synth can write the validation reports to a file 1`] = `
"{
  \\"reports\\": [
    {
      \\"title\\": \\"Validation Report (<__dirname-replaced>/__resources__/validation-plugin@0.0.0)\\",
      \\"violations\\": [
        {
          \\"ruleName\\": \\"Some rule\\",
          \\"recommendation\\": \\"Some recommendation\\",
          \\"violatingConstructs\\": [
            {
              \\"manifestPath\\": \\"dist/chart-c86185a7.k8s.yaml\\",
              \\"resourceName\\": \\"config-map\\",
              \\"locations\\": [
                \\"location1\\",
                \\"location2\\"
              ],
              \\"constructPath\\": \\"Chart/Object\\"
            }
          ],
          \\"fix\\": \\"Some fix\\"
        }
      ],
      \\"summary\\": {
        \\"status\\": \\"success\\",
        \\"plugin\\": \\"<__dirname-replaced>/__resources__/validation-plugin\\",
        \\"version\\": \\"0.0.0\\"
      }
    }
  ]
}"
`;
